Public/Src/Sandbox/MacOs/Interop/Posix/memory.c

```diff
@@ -42,15 +42,58 @@ int GetRamUsageInfo(RamUsageInfo *buffer, long bufferSize)
     return KERN_SUCCESS;
 }
 
+static uint64_t ProcessTreeWorkingSetSize(pid_t pid, const rlim_t max_proc_count, bool *success)
+{
+    uint64_t children_usage = 0;
+
+    pid_t pids[max_proc_count];
+    int child_count = proc_listchildpids(pid, pids, (int) max_proc_count);
+    *success &= child_count >= 0;
+
+    for (int i = 0; (i < child_count) && *success; i++)
+    {
+        int pid = pids[i];
+        rusage_info_current rusage;
+        *success &= proc_pid_rusage(pid, RUSAGE_INFO_CURRENT, (void **)&rusage) == 0;
+        children_usage += rusage.ri_resident_size;
+
+        pid_t childs_pids[max_proc_count];
+        int count = proc_listchildpids(pid, childs_pids, (int) max_proc_count);
+        *success &= count >= 0;
+
+        if (count != 0)
+        {
+            children_usage += ProcessTreeWorkingSetSize(pid, max_proc_count, success);
+        }
+    }
+
+    return children_usage;
+
+}
+
 int GetPeakWorkingSetSize(pid_t pid, uint64_t *buffer)
 {
+    struct rlimit rl;
+    bool success = getrlimit(RLIMIT_NPROC, &rl) == 0;
+
     rusage_info_current rusage;
-    if (proc_pid_rusage(pid, RUSAGE_INFO_CURRENT, (void **)&rusage) != 0)
+    success &= proc_pid_rusage(pid, RUSAGE_INFO_CURRENT, (void **)&rusage) == 0;
```

Comment:
I think it would be better if `ProcessTreeWorkingSetSize` would include `proc_pid_rusage` result for the root pid and all of its children so that you don't have to call it here first.